rumqttd/src/router/routing.rs

```diff
@@ -1193,11 +1204,12 @@ fn append_to_commitlog(
     if publish.payload.is_empty() {
         datalog.remove_from_retained_publishes(topic.to_owned());
     } else if publish.retain {
-        error!("Unexpected: retain field was not unset");
         datalog.insert_to_retained_publishes(publish.clone(), properties.clone(), topic.to_owned());
+        return Ok(None);
     }
 
-    publish.retain = false;
+    // NOTE(swanandx): why we unset the retain field?
+    // publish.retain = false;
```

Comment:
```suggestion

```

Reply:
turns out, it was required. as we do need to append that message to commitlog for existing subscribers!